Add basic functionality of handle:

ifedchankau opened the pull request:
- Add implementation of handle function
- Add functions parseSection and parseOption (w/o implementation)

Closes #18

xferra commented on src/handle.js:
```diff
@@ -1,6 +1,25 @@
+
     let options = [];
+    parseSection(sectionName, help).forEach(section => {
+        let splitTmp = (section.split(/\n[ \t]*(-\S+?)/)).slice(1);
```
Unnecessary `()` and move regex value to a variable, ideally that has a descriptive name what it does.

xferra commented on src/handle.js:
```diff
@@ -1,6 +1,25 @@
+    parseSection(sectionName, help).forEach(section => {
+        let splitTmp = (section.split(/\n[ \t]*(-\S+?)/)).slice(1);
+        let split = [];
+        for (let cnt = splitTmp.length, i = 0; i < cnt; i+=2) {
```
I suggest to isolate it in a function like `splitPairs`.

xferra commented on src/handle.js:
```diff
@@ -1,6 +1,25 @@
+        let split = [];
+        for (let cnt = splitTmp.length, i = 0; i < cnt; i+=2) {
+            if (splitTmp !== undefined) {
+                split.push(splitTmp[i] + splitTmp[i + 1]);
```
Probably `[ splitTmp[i] , splitTmp[i + 1] ]`?

xferra replied:
I'm trying to follow the logic here, split then concat back, what is the idea?

ifedchankau replied:
Sorry, i forgot about set flag PREG_SPLIT_DELIM_CAPTURE of preg_split php function when transating it to js. If this flag is set, parenthesized expression in the delimiter pattern will be captured and returned as well. This is also an answer on your question about parenthes in regular expression.

For example, with reg exp /\n[ \t]*(-\S+?)/ string
"-c, --config          configuration file                                                                                
  -e, --exclude         exclude globs from path expansion"                                                              
will be sliced into 4 strings:  
* "-c" 
* ", --config          configuration file"
* "-e" 
* ", --exclude         exclude globs from path expansion"                                                           

So, its needed to concate 1 and 2, 3 and 4 strings later

Maybe you know, what is js alternative for PREG_SPLIT_DELIM_CAPTURE flag or can suggest another solution of this?

xferra replied:
Like this one?
https://searchcode.com/codesearch/view/14390265/#

There is no alternative out of the box.

xferra commented on src/handle.js:
```diff
@@ -1,6 +1,25 @@
+                split.push(splitTmp[i]);
+            }
+        }
+        split.forEach(option => {
```
Just FYI, alternative variant:
```javascript
split.filter((x) => x. startsWith('i')).map((x) => parseOption(section))
```

xferra commented on src/parseOption.js:
```diff
@@ -0,0 +1,8 @@
+const parseOption = (optionDescription) => {
```
I hope there is no need to create separate modules and files for each function. You can export an object as well:
```javascript
const parseOption = () => {};
const parseSection = () => {};

module.exports = {
   parseOption: parseOption,
   parseSection: parseSection
};

```

xferra replied:
+Consider using [classes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes), JS supports not the only [procedural paradigm](https://en.wikipedia.org/wiki/Procedural_programming).

ifedchankau committed:
feat: add handle function

- add implementation of handle function
- add functions parseSection, splitSection, parseOption (w/o implementation)

Closes #18

xferra reviewed (approved):
+1

romanrostislavovich commented:
:tada: This PR is included in version 1.0.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@linterhub/usage-parser)
- [GitHub release](https://github.com/linterhub/usage-parser/releases/tag/v1.0.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
